torch/nn/functional.py

```diff
@@ -3953,3 +3953,188 @@ def multi_head_attention_forward(query,                           # type: Tensor
+                seq, num_heads, in_proj_weight, in_proj_bias=in_proj_bias)
+    seq_len, bsz, _ = seq.size()
+    proj_dim = in_proj_weight.size(0)
+    assert proj_dim % num_heads == 0, "projection dimension must be divisible by num_heads"
+    head_dim = proj_dim // num_heads
+
+    q = linear(seq, in_proj_weight, in_proj_bias)
+    # Shape of q: (S, N, P)
+    q = q.reshape(seq_len, bsz * num_heads, head_dim).transpose(0, 1)
+    return q
+
+
+def scaled_dot_product_attention(q,                         # type: Tensor
+                                 k,                         # type: Tensor
+                                 v,                         # type: Tensor
+                                 num_heads,                 # type: int
+                                 add_zero_attn,             # type: bool
+                                 dropout_p,                 # type: float
+                                 training=True,             # type: bool
+                                 key_padding_mask=None,     # type: Optional[Tensor]
+                                 attn_mask=None,            # type: Optional[Tensor]
+                                 ):
+    # type: (...) -> Tuple[Tensor, Tensor]
+    r"""Uses a scaled dot product with the projected key-value pair to update
+    the projected query.
+
+    Args:
+        q (Tensor): Projected query
+        k (Tensor): Projected key
+        v (Tensor): Projected value
+        num_heads (int): Number of parallel attention heads.
+        add_zero_attn (bool): Add a new batch of zeros to the projected key and
+            value sequences at dimension 1.
+        dropout_p (float): Probability of an element will be zeroed.
+        training (bool): Apply dropout if ``training=True``
+        key_padding_mask (Tensor, optional): Specified padding elements in the
+            key will be ignored by the attention. This is a binary mask. When
+            the value is True, the corresponding value on the attention layer
+            will be set to :math:`-\inf`.
+        attn_mask (Tensor, optional): 2D or 3D mask that prevents attention to
+            certain positions. This is an additive mask (i.e. the values will
+            be added to the attention layer). A 2D mask will be broadcasted for
+            all the batches while a 3D mask allows to specify a different mask
+            for the entries of each batch.
+
+    Shape:
+        - q: :math:`(N * H, L, P / H)`
+        - k: :math:`(N * H, S, P / H)`
+        - v: :math:`(N * H, S, P / H)`
+        - key_padding_mask: :math:`(N, S)`
+        - attn_mask: :math:`(L, S)` or :math:`(N * H, L, S)`
+        - Output: :math:`(N * H, L, P / H)`, :math:`(N * H, L, S)`
+        where L is the target length, S is the source length, H is the number
+        of attention heads, N is the batch size, and P is the projection
+        dimension.
+
+    """
+    if not torch.jit.is_scripting():
```

Comment:
Why is this part needed?

Reply:
I have no idea. It's included with the rest, and I assumed it was part of the `jit` package which I'm not fully familiar with.